pkg/gatewayserver/io/ws/lbslns/upstream.go

```diff
@@ -512,28 +522,35 @@ func (f *lbsLNS) HandleUp(ctx context.Context, raw []byte, ids ttnpb.GatewayIden
 			ID: int32(updf.UpInfo.XTime >> 48),
 		}
 		session.DataMu.Unlock()
-		recordTime(updf.RefTime, updf.UpInfo.XTime, receivedAt)
+		recordTime(0, updf.UpInfo.XTime, receivedAt)
 
 	case TypeUpstreamTxConfirmation:
 		var txConf TxConfirmation
 		if err := json.Unmarshal(raw, &txConf); err != nil {
 			return nil, err
 		}
-		txAck := txConf.ToTxAck(ctx, f.tokens, receivedAt)
+		txAck, delta := txConf.ToTxAck(ctx, f.tokens)
 		if txAck == nil {
 			break
 		}
 		if err := conn.HandleTxAck(txAck); err != nil {
 			logger.WithError(err).Warn("Failed to handle tx ack message")
 			return nil, err
 		}
+
 		session := ws.SessionFromContext(ctx)
 		session.DataMu.Lock()
 		session.Data = State{
 			ID: int32(txConf.XTime >> 48),
 		}
 		session.DataMu.Unlock()
-		recordTime(txConf.RefTime, txConf.XTime, receivedAt)
+
+		if delta > f.maxRoundTripDelay {
```

Comment:
The server time minus `RefTime` _is_ the delta.

`RefTime` is in the server time domain as it's sent as `MuxTime`.

Why are downlink tokens used for this?